### ISSUE TITLE
Using valid ws4redis URL; Group name must be a valid unicode string containing only alphanumerics, hyphens, or periods

### ISSUE THREAD

**Issue by ThaJay** [django/channels#522]
As a result of ws4redis not working with Channels installed (it calls runserver for its uwsgi but gets Daphne, see link at the bottom), I want to make our old URL work as a group for as long as we need it, until our clients have all updated their mobile apps and we don't need the old implementation to work any more.

This is the url in question:
`wss://www.website.com/ws/GuestProduct?subscribe-broadcast&publish-group&echo`

and this is the error it throws:
```
Traceback (most recent call last):
  File "/srv/platform/production/env/lib/python3.4/site-packages/channels/worker.py", line 119, in run
    consumer(message, **kwargs)
  File "/srv/platform/production/src/project/consumers.py", line 44, in ws_disconnect
    Group(get_group(message)).discard(message.reply_channel)
  File "/srv/platform/production/env/lib/python3.4/site-packages/channels/channel.py", line 75, in discard
    self.channel_layer.group_discard(self.name, channel)
  File "/srv/platform/production/env/lib/python3.4/site-packages/asgi_redis/core.py", line 233, in group_discard
    assert self.valid_group_name(group), "Group name not valid"
  File "/srv/platform/production/env/lib/python3.4/site-packages/asgiref/base_layer.py", line 114, in valid_group_name
    raise TypeError("Group name must be a valid unicode string containing only alphanumerics, hyphens, or periods.")
TypeError: Group name must be a valid unicode string containing only alphanumerics, hyphens, or periods.
```
Because I learned earlier that underscores are actually allowed, it seems like it throws on the `&`, even though it's part of the querystring, while I haven't been able to read querystrings myself in `ws_connect`.

our group name at the moment:
(I'm going to look into Daphne's default url prefix setting later)
```
def get_group(message):
    return message.content['path'].strip("/").replace("ws/", "", 1)
```

some additional info:
`Ubuntu 14.04.1 LTS`

```
Django==1.10.3
Twisted==16.6.0
asgi-redis==1.0.0
channels==1.0.2
daphne==1.0.1
django-websocket-redis==0.4.7
uWSGI==2.0.13
```
http://stackoverflow.com/questions/42002321/using-ws4redis-with-channels-installed

**Comment by andrewgodwin** [comment]
I'm not exactly sure what the bug you're reporting is - the error message tells you the problem and you haven't actually said what the different behaviour you're expecting is.

What do you want us to do?

**Comment by ThaJay** [comment]
Its in the first sentence. Either I need this URL to work on channels, or I need ws4redis to work with channels installed, for backwards compatibility with our old implementation.

Why is it giving me this error? is it because I'm using an ampersand even if it's in the querystring?
If so I would like a way to use it anyway, as I'm not able to change this url on old versions of our app that are still in use.

On the other hand, if this URL restriction can't or won't be fixed, I need to file a different issue related to the stackoverflow question I linked in my opening post, about how runserver defaults to Daphne once channels is installed.

Another option is to file an issue on ws4redis because they are the ones calling runserver instead of uwsgi directly. If that also doesn't pan out we will have to find a way to get what we want without channels.

**Comment by andrewgodwin** [comment]
If you can tell me what the channel name that you're trying to send on is, then I can tell you the error. Just stick a `print()` around `get_group` and see what it says.

**Comment by ThaJay** [comment]
I tried to do that but it did not show so I thought for some reason it got rejected before. So I left it for the rest of the day and focussed on other things.
This morning I finally found out what the problem was. It's actually a bug in Supervisor, the stdout log (where a print statement apparently appears) does not update until you stop the program (the worker in this case). I did not think to check this before because I only saw data appear in the stderr log file, which did actually update as soon as the worker logged something.

Today I went back to 1 worker and used `multitail` to log both files in a pane at the same time and my `print` appeared after restarting the worker.
So finally seeing a sensible log, channels does actually eat the querystring and I can just have a group called `GuestProduct` like I expected.

This means we can actually use channels to have our old apps still work so we will just remove ws4redis and use channels for all our websockets.

Nice! When this is done we can start to think about ways to make our dashboard feel more alive.

Thank you so much for your patience and I'm sorry for wasting your time, thinking there was something going wrong in the channels code. Coding is one thing, but it's still not easy for me to understand every piece of the puzzle all the time. Especially when some of them don't work as expected.
Now I can't wait to start using channels more.
